caikit_nlp/data_model/embedding_vectors.py

```diff
@@ -0,0 +1,188 @@
+error = error_handler.get(log)
+
+
+@dataobject(package="caikit_data_model.caikit_nlp")
+class FloatValue(DataObjectBase):
+    float_value: Union[
+        Annotated[float, OneofField("data_float"), FieldNumber(1)],
+        Annotated[np.float32, OneofField("data_float32"), FieldNumber(2)],
+        Annotated[np.float64, OneofField("data_float64"), FieldNumber(3)],
+    ]
+
+
+@dataobject(package="caikit_data_model.caikit_nlp")
+class Vector1D(DataObjectBase):
+    """Data representation for a 1 dimension vector.
+    NOTE: This only support float types
+    """
+
+    data: List[FloatValue]
+
+    def _convert_np_to_list(self, values):
+        return values.tolist()
+
+    def _convert_to_floatvalue(self, values):
+
+        if isinstance(values, np.ndarray):
+            if values.dtype == np.float32:
+                return [FloatValue(data_float32=val).to_proto() for val in values]
+            if values.dtype == np.float64:
+                return [FloatValue(data_float64=val).to_proto() for val in values]
+        return [FloatValue(val).to_proto() for val in values]
+
+    def to_dict(self) -> dict:
+
+        # If data is in np.ndarray format, convert it to python list
+        if isinstance(self.data, np.ndarray):
+            return {"data": self._convert_np_to_list(self.data)}
+
+        return {"data": self.data}
+
+    def fill_proto(self, proto):
+        data_proto = getattr(proto, "data")
+        data_proto.extend(self._convert_to_floatvalue(self.data))
+
+        return proto
+
+    @classmethod
+    def from_proto(cls, proto):
+        data = []
+        for datum in proto.data:
+            value_type = FloatValue.from_proto(datum).which_oneof("float_value")
+            data.append(getattr(datum, value_type))
+        return cls(**{"data": data})
+
+    @classmethod
+    def from_json(cls, json_str):
+        """Function to convert json_str to DataBase object
+
+        NOTE: Since JSON doesn't support variety of float formats
+              we would only be able to convert all the values to python floats here.
+              So there would be a loss of information when converting to and from JSON
+
+        Args:
+            json_str: str or dict
+                A stringified JSON specification/dict of the data_model
+
+        Returns:
+            caikit.core.data_model.DataBase
+                A DataBase object.
+        """
+        error.type_check("<NLP97633962E>", dict, str, json_str=json_str)
+        if isinstance(json_str, str):
+            json_str = json.loads(json_str)
+
+        error.value_check(
+            "<NLP49158301E>", "data" in json_str, "invalid json_str provided!"
+        )
+
+        data = []
+        for datum in json_str["data"]:
+            data.append(datum)
+        return cls(data=data)
+
+
+@dataobject(package="caikit_data_model.caikit_nlp")
+class EmbeddingResult(DataObjectBase):
+    """Data representation for an embedding matrix holding 2D vectors"""
+
+    data: List[Vector1D]
```

Comment:
Hm, interesting. Even if we go with the above where all elements of the sequence need to be of a common type, you could have sequences here that are (a) different type and (b) different length. Could that cause any problems?

Reply:
Mismatch in length of individual list can certainly be the case with this approach. So essentially 2 vectors may not have equal length. But for now, I don't think we need to guarantee that at data model level. Like if the data is in python `List[List[]]`, which is a type we are supporting here, then that is anyways a possibility. 🤔